torch_xla/distributed/xrt_init.py

```diff
@@ -0,0 +1,241 @@
+        else:
+            address = _TCP_STORE.get(key).decode('UTF-8')
+        if total_nodes == 1:
+            xrt_tpu_config.append(f'localservice;{i};{address}')
+        else:
+            xrt_tpu_config.append(f'c_localservice;{i};{address}')
+
+        if rank == 0:
+            os.environ[xenv.TPU_CONFIG] = '|'.join(xrt_tpu_config)
+            os.environ[xenv.TPU_NUM_DEVICES] = str(local_world_size)
+
+    os.environ[
+        xenv.
+        LOCAL_WORKER] = f'localservice:{group_rank}' if total_nodes == 1 else f'c_localservice:{group_rank}'
+    os.environ[xenv.WORLD_SIZE] = str(world_size)
+    os.environ[xenv.HOST_WORLD_SIZE] = str(total_nodes)
+    os.environ[xenv.ORDINAL] = str(rank)
+    os.environ[xenv.LOCAL_ORDINAL] = str(local_rank)
+    os.environ[xenv.MP_DEVICE] = f'TPU:{rank}'
+    if not os.environ.get('TF_GRPC_DEFAULT_OPTIONS', None):
+        os.environ['TF_GRPC_DEFAULT_OPTIONS'] = (
+            'grpc.keepalive_time_ms=60000,grpc.keepalive_timeout_ms=14400000,'
+            'grpc.http2.max_pings_without_data=0,grpc.http2.min_ping_interval_without_data_ms=300000'
+        )
+    # We don't want torch_xla to start the local server internally.
+    # We are starting the xrt server by ourselves
+    os.environ['XRT_START_LOCAL_SERVER'] = '0'
+
+    return tpu_config_port
+
+
+def _set_neuron_envs(rank, world_size, local_world_size):
+    os.environ["NEURON_USE_LOAD_COLLECTIVES"] = '1'
+    os.environ['NEURON_GLOBAL_DEVICE_ID'] = str(rank)
+    os.environ['NEURON_GLOBAL_DEVICE_COUNT'] = str(world_size)
+    if not os.environ.get('NEURON_RT_VISIBLE_CORES', None):
+        os.environ['NEURON_RT_VISIBLE_CORES'] = ','.join(
+            [str(i) for i in range(local_world_size)])
+
+
+def _setup_nccl_service(rank):
+    # Set up NCCL COMM ID required for NCCL communicator IDs
+    address = _get_address_from_store('nccl_info', rank)
+    os.environ['NEURON_RT_ROOT_COMM_ID'] = address
+
+
+def set_xrt_envs(dev_kind, world_size, rank, local_rank):
+    # Set up all the XRT specific env variables, adapted from xmp.spawn()
+    os.environ[xenv.WORLD_SIZE] = str(world_size)
+    os.environ[xenv.ORDINAL] = str(rank)
+    os.environ[xenv.LOCAL_ORDINAL] = str(local_rank)
+    os.environ[xenv.LOCAL_WORKER] = 'localservice:' + str(rank)
+
+    os.environ[xenv.MP_DEVICE] = f'GPU:{rank}'
+    gpus_to_use = os.environ.get('CUDA_VISIBLE_DEVICES')
+    if gpus_to_use is not None:
+        # If gpu devices are set by a scheduling entity (eg. SLURM) we index into
+        # comma separated string containing numbered gpu devies
+        gpus_to_use_list = gpus_to_use.split(',')
+        os.environ['CUDA_VISIBLE_DEVICES'] = gpus_to_use_list[local_rank]
+    else:
+        # If no explicit visible devices are provided, local_rank is used to identify
+        # the gpu used by this process
+        os.environ['CUDA_VISIBLE_DEVICES'] = str(local_rank)
+
+
+def _init_xrt_context(master_addr=None, master_port=None, store=None):
+    '''
```

Comment:
nit: Can you update this docstring to match the format used in other files? e.g. https://github.com/pytorch/xla/blob/590dee5d08ff4cf3ba72eab306b470e22bd08e16/torch_xla/core/xla_model.py#L187-L197

Reply:
Sure, no problem!